### ISSUE TITLE
Loss does not converge!

### ISSUE THREAD

**Issue by gaoj21** [ChelsieLei/HOLa#3]
➜  HOLa git:(master) ✗ bash scripts/hico_vitB.sh

Namespace(lr_head=0.001, lr_vit=0.001, batch_size=16, weight_decay=0.0001, epochs=12, lr_drop=10, clip_max_norm=0.1, backbone='resnet50', dilation=False, position_embedding='sine', repr_dim=512, hidden_dim=256, enc_layers=6, dec_layers=6, dim_feedforward=2048, dropout=0.1, nheads=8, num_queries=100, pre_norm=False, aux_loss=True, set_cost_class=1, set_cost_bbox=5, set_cost_giou=2, bbox_loss_coef=5, giou_loss_coef=2, eos_coef=0.1, alpha=0.5, gamma=0.2, dataset='hicodet', partitions=['train2015', 'test2015'], num_workers=2, data_root='./hicodet', device='cuda', port='1235', seed=66, pretrained='checkpoints/detr-r50-hicodet.pth', resume='', output_dir='checkpoints/hico_HO_adpt_uv_512_clipbase/hola', print_interval=500, world_size=1, eval=False, cache=False, sanity=False, box_score_thresh=0.2, fg_iou_thresh=0.5, min_instances=3, max_instances=15, visual_mode='vit', clip_dir_vit='checkpoints/pretrained_CLIP/ViT-B-16.pt', clip_visual_layers_vit=24, clip_visual_output_dim_vit=768, clip_visual_input_resolution_vit=336, clip_visual_width_vit=1024, clip_visual_patch_size_vit=14, clip_text_transformer_width_vit=768, clip_text_transformer_heads_vit=12, clip_text_transformer_layers_vit=12, clip_text_context_length_vit=77, use_insadapter=True, use_mean=False, logits_type='HO+U', num_shot=2, file1='hicodet_pkl_files/union_embeddings_cachemodel_crop_padding_zeros_vitb16.p', prior_type='cbe', training_set_ratio=1.0, frozen_classifier=None, zs=True, hyper_lambda=2.8, use_weight_pred=False, zs_type='unseen_verb', vis_tor=1.0, adapter_num_layers=1, N_CTX=24, CSC=False, CTX_INIT='', CLASS_TOKEN_POSITION='end', use_templates=False, feat_mask_type=0, num_classes=117, prior_method=0, vis_prompt_num=50, box_proj=0, adapter_pos='all', use_multi_hot=True, label_learning=False, label_choice='random', repeat_factor_sampling=False, mask_loss_coef=1, dice_loss_coef=1, cls_loss_coef=2, focal_alpha=0.25, vis_img=False, vis_img_path='pred_annotation', local_rank=0, resume_part='', fix_mem=False, llmtxt=True, img_align=True, semloss_weight=150, self_adapt=True, norm_pred=False, wo_unseen_pred=True, basis_feat_enable=True, seperate_ho=1, basis_feat_init='pca', unique_basis_weights=True, disentangle_basis=True, basis_num=100, ao_sep_basis=True, act_txtdecrip=True, sep_frac=3, basis_constraint='quadratic', basis_feat_constraint='none', fix_act_w=False, HOI_train_w_b='w', no_act_constraint=False, kl_t=0.1, recon_ratio_pca=71.0, wo_sparsity=False, pt_learn=0, pt_lyr=[1, 9], semloss=True, ho_pair_pt=True, ho_pair_prior=1, pt_init='pos+detr', pred_type='ho+u+l', pt_attn='uniform')
ViT-B/16
ViT-B/16
[INFO]: num_classes 117
Load weights for the object detector from checkpoints/detr-r50-hicodet.pth
CLIP image resolution 224
CLIP layer number 12
[INFO] missing_keys: []
[INFO] unexpected_keys: []
low rank:  71
=> Rank 0: start from a randomly initialised model
number of leanable params: 4024922
number of all params: 195308048
Epoch [1/12], Iter. [0500/2222], Loss: 53.1089, Time[Data/Iter.]: [39.75s/1164.78s]
Epoch [1/12], Iter. [1000/2222], Loss: 53.6479, Time[Data/Iter.]: [38.81s/1166.66s]
tensor(0, device='cuda:0') 15
Epoch [1/12], Iter. [1500/2222], Loss: 56.1481, Time[Data/Iter.]: [38.97s/1175.36s]
tensor(0, device='cuda:0') 11
Epoch [1/12], Iter. [2000/2222], Loss: 57.3153, Time[Data/Iter.]: [38.84s/1174.58s]
Epoch [2/12], Iter. [0278/2222], Loss: 58.1829, Time[Data/Iter.]: [40.16s/1180.02s]
tensor(0, device='cuda:0') 9
Epoch [2/12], Iter. [0778/2222], Loss: 58.8071, Time[Data/Iter.]: [38.36s/1182.41s]
tensor(0, device='cuda:0') 15

**Comment by ChelsieLei** [comment]
Hi @gaoj21 

Thanks for your interest. What we use for training is 4 A5000 GPUs and each of them has a 16 batch. I think your divergence can be caused by the small batch size. You can give it a try with a 64 batch size in total and re-train again. 

Best,
Chelsie

**Comment by gaoj21** [comment]
(pvic) ➜  HOLa git:(master) ✗ bash scripts/hico_vitB.sh\

Namespace(lr_head=0.001, lr_vit=0.001, batch_size=16, weight_decay=0.0001, epochs=12, lr_drop=10, clip_max_norm=0.1, backbone='resnet50', dilation=False, position_embedding='sine', repr_dim=512, hidden_dim=256, enc_layers=6, dec_layers=6, dim_feedforward=2048, dropout=0.1, nheads=8, num_queries=100, pre_norm=False, aux_loss=True, set_cost_class=1, set_cost_bbox=5, set_cost_giou=2, bbox_loss_coef=5, giou_loss_coef=2, eos_coef=0.1, alpha=0.5, gamma=0.2, dataset='hicodet', partitions=['train2015', 'test2015'], num_workers=2, data_root='./hicodet', device='cuda', port='1235', seed=66, pretrained='checkpoints/detr-r50-hicodet.pth', resume='', output_dir='checkpoints/hico_HO_adpt_uv_512_clipbase/hola', print_interval=100, world_size=3, eval=False, cache=False, sanity=False, box_score_thresh=0.2, fg_iou_thresh=0.5, min_instances=3, max_instances=15, visual_mode='vit', clip_dir_vit='checkpoints/pretrained_CLIP/ViT-B-16.pt', clip_visual_layers_vit=24, clip_visual_output_dim_vit=768, clip_visual_input_resolution_vit=336, clip_visual_width_vit=1024, clip_visual_patch_size_vit=14, clip_text_transformer_width_vit=768, clip_text_transformer_heads_vit=12, clip_text_transformer_layers_vit=12, clip_text_context_length_vit=77, use_insadapter=True, use_mean=False, logits_type='HO+U', num_shot=2, file1='hicodet_pkl_files/union_embeddings_cachemodel_crop_padding_zeros_vitb16.p', prior_type='cbe', training_set_ratio=1.0, frozen_classifier=None, zs=True, hyper_lambda=2.8, use_weight_pred=False, zs_type='unseen_verb', vis_tor=1.0, adapter_num_layers=1, N_CTX=24, CSC=False, CTX_INIT='', CLASS_TOKEN_POSITION='end', use_templates=False, feat_mask_type=0, num_classes=117, prior_method=0, vis_prompt_num=50, box_proj=0, adapter_pos='all', use_multi_hot=True, label_learning=False, label_choice='random', repeat_factor_sampling=False, mask_loss_coef=1, dice_loss_coef=1, cls_loss_coef=2, focal_alpha=0.25, vis_img=False, vis_img_path='pred_annotation', local_rank=0, resume_part='', fix_mem=False, llmtxt=True, img_align=True, semloss_weight=150, self_adapt=True, norm_pred=False, wo_unseen_pred=True, basis_feat_enable=True, seperate_ho=1, basis_feat_init='pca', unique_basis_weights=True, disentangle_basis=True, basis_num=100, ao_sep_basis=True, act_txtdecrip=True, sep_frac=3, basis_constraint='quadratic', basis_feat_constraint='none', fix_act_w=False, HOI_train_w_b='w', no_act_constraint=False, kl_t=0.1, recon_ratio_pca=71.0, wo_sparsity=False, pt_learn=0, pt_lyr=[1, 9], semloss=True, ho_pair_pt=True, ho_pair_prior=1, pt_init='pos+detr', pred_type='ho+u+l', pt_attn='uniform')
ViT-B/16
ViT-B/16
ViT-B/16
ViT-B/16
ViT-B/16
ViT-B/16
[INFO]: num_classes 117
[INFO]: num_classes 117
[INFO]: num_classes 117
Load weights for the object detector from checkpoints/detr-r50-hicodet.pth
CLIP image resolution 224
CLIP layer number 12
CLIP image resolution 224
CLIP layer number 12
CLIP image resolution 224
CLIP layer number 12
[INFO] missing_keys: []
[INFO] unexpected_keys: []
[INFO] missing_keys: []
[INFO] unexpected_keys: []
[INFO] missing_keys: []
[INFO] unexpected_keys: []
low rank:  71
=> Rank 2: start from a randomly initialised model
low rank:  71
=> Rank 1: start from a randomly initialised model
low rank:  71
=> Rank 0: start from a randomly initialised model
number of leanable params: 4024922
number of leanable params: 4024922
number of leanable params: 4024922
number of all params: 195308048
number of all params: 195308048
number of all params: 195308048
Epoch [1/12], Iter. [100/740], Loss: 67.0282, Time[Data/Iter.]: [15.46s/273.25s]
tensor(0, device='cuda:1') 15
tensor(0, device='cuda:2') 15
Epoch [1/12], Iter. [200/740], Loss: 48.3657, Time[Data/Iter.]: [8.19s/263.27s]
Epoch [1/12], Iter. [300/740], Loss: 49.1840, Time[Data/Iter.]: [8.03s/265.79s]
Epoch [1/12], Iter. [400/740], Loss: 50.8331, Time[Data/Iter.]: [8.08s/266.66s]
Epoch [1/12], Iter. [500/740], Loss: 51.9927, Time[Data/Iter.]: [8.25s/266.92s]
Epoch [1/12], Iter. [600/740], Loss: 53.3736, Time[Data/Iter.]: [7.98s/265.53s]
Epoch [1/12], Iter. [700/740], Loss: 54.5984, Time[Data/Iter.]: [8.23s/267.43s]
Epoch [2/12], Iter. [060/740], Loss: 55.5741, Time[Data/Iter.]: [15.37s/275.32s]
Epoch [2/12], Iter. [160/740], Loss: 56.3596, Time[Data/Iter.]: [8.14s/269.19s]
tensor(0, device='cuda:1') 15
Epoch [2/12], Iter. [260/740], Loss: 56.8178, Time[Data/Iter.]: [8.05s/267.05s]
Epoch [2/12], Iter. [360/740], Loss: 57.3750, Time[Data/Iter.]: [8.07s/266.23s]
Epoch [2/12], Iter. [460/740], Loss: 57.6609, Time[Data/Iter.]: [8.17s/268.12s]
Epoch [2/12], Iter. [560/740], Loss: 57.5395, Time[Data/Iter.]: [8.12s/266.28s]


I used 3 A5000 GPUs. It was still not converging. The batch size may not be the cause of the divergence. Could you please check if you've updated the correct code?

**Comment by ChelsieLei** [comment]
Hi @gaoj21 

Thanks for your information. I found that there is indeed a bug in hola_model.py when calculating the reconstruction loss. 
I fixed the problem and updated the file hola_model.py
I checked that this version can work without problem. You can give it a try, thanks.

**Comment by gaoj21** [comment]
Thanks for your update. The code works fine so far.

**Comment by ChelsieLei** [comment]
No problem. Thanks to point out this issue.
